File: testsystem-backend/src/main/java/com/underfit/testsystembackend/service/TestSystemService.java
package com.underfit.testsystembackend.service;

import com.underfit.testsystembackend.dto.*;

import java.util.List;

public interface TestSystemService {
    /**
     * Получение всех тестов системы
     *
     * @return тесты
     */
    List<TestDto> getTests();

    /**
     * Получение теста по id, связанных вопросов и вариантов ответов к ним
     *
     * @param id теста
     * @return Тест и вопросы, и варианты ответа к нему
     */
    TestWithQuestionDto getTestById(Long id);

    /**
     * Получение вопроса по id вопроса и ответов к нему
     *
     * @param id вопроса
     * @return вопрос с вариантами ответа
     */
    QuestionWithOptionDto getQuestionById(Long id);

    /**
     * Сохранение результата теста для пользователя
     *
     * @param createResultDto Результат пользователя по тесту
     * @return Сохраненный результат
     */
    ResultDto createResult(CreateResultDto createResultDto);

    /**
     * Получение критерия оценивания за тест по баллам, которые набрал за тест пользователь
     *
     * @param testId тест
     * @param score  набранные баллы
     * @return критерий оценивания (результат)
     */
    AssessmentDto getAssessmentByTestIdAndScore(Long testId, Integer score);

    /**
     * Принятие результата теста или нет
     *
     * @param resultId Результат теста
     * @param accept   Принять/отклонить
     */
    void acceptResult(Long resultId, Boolean accept);

    /**
     * Сохранение ответов пользователя на вопросы теста
     * @param createAnswers ответы
     * @return результат сохранения ответов
     */
    List<AnswerDto> createAnswers(List<CreateAnswerDto> createAnswers);

    /**
     * Получение ответов на тест для пользователя(по resultId)
     * @param resultId Результат теста
     * @return Ответы
     */
    List<AnswerDto> getAnswersByResultId(Long resultId);
}

File: testsystem-frontend/src/app/services/test-system.service.ts
import { Injectable } from '@angular/core';
import {environment} from "../../environments/environment";
import {HttpClient, HttpHeaders} from "@angular/common/http";
import {ResultDto} from "../models/result-dto";
import {TestDto} from "../models/test-dto";
import {TestWithQuestionDto} from "../models/test-with-question-dto";
import {QuestionWithOptionDto} from "../models/question-with-option-dto";
import {CreateResultDto} from "../models/create-result-dto";
import {AssessmentDto} from "../models/assessment-dto";
import {CreateAnswerDto} from "../models/create-answer-dto";
import {AnswerDto} from "../models/answer-dto";

const API = '/api/v1/test-system';
const HOST = environment.apiUrl;

const httpOptions = {
  headers: new HttpHeaders({'Content-Type': 'application/json'})
};

@Injectable({
  providedIn: 'root'
})
export class TestSystemService {

  constructor(private http: HttpClient) { }

  getTests() {
    return this.http.get<Array<TestDto>>(HOST + API + '/tests');
  }

  getTestById(id: number) {
    return this.http.get<TestWithQuestionDto>(HOST + API + '/test' + '/' + id);
  }

  getQuestionById(id: number) {
    return this.http.get<QuestionWithOptionDto>(HOST + API + '/question' + '/' + id);
  }

  createResult(result: CreateResultDto) {
    return this.http.post<ResultDto>(HOST + API + '/create/result', result, httpOptions);
  }

  getAssessmentByTestIdAndScore(testId: number, score: number) {
    return this.http.get<AssessmentDto>(HOST + API + '/assessment' + '/' + testId + '/' + score, httpOptions);
  }

  acceptResult(resultId: number, accept: boolean) {
    return this.http.get<any>(HOST + API + '/accept/result' + '/' + resultId + '/' + accept);
  }

  createAnswers(answers: Array<CreateAnswerDto>) {
    return this.http.post<Array<AnswerDto>>(HOST + API + '/create/answers', answers, httpOptions);
  }

  getAnswersByResultId(resultId: number) {
    return this.http.get<Array<AnswerDto>>(HOST + API + '/answers/result' + '/' + resultId, httpOptions);
  }
}
